### PR TITLE
test: test the >10 UTXO case for output groups

### PR DESCRIPTION
**Pull request by kallewoof**
#17603 notes behavior where `avoidreuse` will only spend 1 of the UTXO:s ~and mark all of them as spent~. This is ~correct~ currently expected behavior, but the reasoning is non-obvious: the user creates 31 outputs, and the code groups these into 4 groups, 3x10 and 1x1. It then ends up using the 1x1 group because it is smaller => fewer fees.

This PR adds a test that explicitly checks this case, but in a more direct manner; it creates 30 outputs of 1 btc each, all sending to the same destination, and the receiver then sends 11 btc back. From [10, 10, 10] coin select will pick 2, leave 1 alone, and create 11 out/9 change. The user will end up with ~~10 btc used, 9 btc unused.~~ 19 btc unused.*

(* this is because the system is currently set up to mark things as they come into the wallet; already existing stuff is only marked if the user does a rescan, which the caveat states when enabling the flag)

### CODE REVIEW

File: test/functional/wallet_avoidreuse.py [hunk 1]
```diff
@@ -240,5 +242,37 @@ def test_fund_send_fund_send(self):
...
         assert_approx(self.nodes[1].getbalance(avoid_reuse=False), 11, 0.001)
 
+    def test_many_utxos_one_dest(self):
+        '''
+        Test the case where [1] generates a new address A, then
+        [0] sends 1 BTC to A 30 times.
+        [1] spends 11 BTC from A
+        20 of the outputs in A should be used, leaving 10 intact.
+        Those 10 should all be marked used.
+        A single output of 9 btc should be change back to self, and should not be marked used.
```
Comment (jonatack) [inline test/functional/wallet_avoidreuse.py]: nits: s/change/changed/, and s/self/itself/ ?

File: test/functional/wallet_avoidreuse.py [hunk 2]
```diff
@@ -240,5 +242,37 @@ def test_fund_send_fund_send(self):
...
+        assert_balances(self.nodes[1], mine={"used": 0, "trusted": 30})
+
+        self.nodes[1].sendtoaddress(retaddr, 11)
+        self.nodes[0].generate(1)
+        self.sync_all()
+
+        # listunspent should show 1 unused 9 btc output, and 10 used 1 btc outputs
+        assert_unspent(self.nodes[1], total_count=11, total_sum=19, reused_supported=True, reused_count=10)
+        # getbalances should show 10 used, 9 btc trusted
+        assert_balances(self.nodes[1], mine={"used": 10, "trusted": 9})
```
Comment (jonatack) [inline test/functional/wallet_avoidreuse.py]: This assert is failing for me with `AssertionError: 0E-8 < [9.999..10.001]`.

**Review by jonatack — COMMENTED** [review]
Thanks for adding the test. The last assertion fails for me with 19 btc trusted instead of the expected 10 used, 9 trusted:
```python
         # getbalances should show 10 used, 9 btc trusted
+        import pprint
+        pprint.pprint(self.nodes[0].getbalances())
+        pprint.pprint(self.nodes[1].getbalances())
         assert_balances(self.nodes[1], mine={"used": 10, "trusted": 9})
```
```
2019-12-31T09:24:55.314000Z TestFramework (INFO): Test many utxos one destination
{'mine': {'immature': Decimal('2262.50308860'),
          'trusted': Decimal('5419.99691140'),
          'untrusted_pending': Decimal('11.00000000')}}
{'mine': {'immature': Decimal('0E-8'),
          'trusted': Decimal('18.99971440'),
          'untrusted_pending': Decimal('0E-8'),
          'used': Decimal('0E-8')}}
2019-12-31T09:25:01.212000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 112, in main
    self.run_test()
  File "test/functional/wallet_avoidreuse.py", line 91, in run_test
    self.test_many_utxos_one_dest()
  File "test/functional/wallet_avoidreuse.py", line 278, in test_many_utxos_one_dest
    assert_balances(self.nodes[1], mine={"used": 10, "trusted": 9})
  File "test/functional/wallet_avoidreuse.py", line 64, in assert_balances
    assert_approx(got[k], v, 0.001)
  File "/home/jon/projects/bitcoin/bitcoin/test/functional/test_framework/util.py", line 31, in assert_approx
    raise AssertionError("%s < [%s..%s]" % (str(v), str(vexp - vspan), str(vexp + vspan)))
AssertionError: 0E-8 < [9.999..10.001]
```

**Comment by kallewoof** [comment]
@jonatack I spoke^Wpushed too soon (I must have been absent-minded as I saw that error locally before pushing).

File: test/functional/wallet_avoidreuse.py (hunk 1 above)
Comment (kallewoof) [reply test/functional/wallet_avoidreuse.py]: It's change, as in money given back, so I think it's correct as it is, but I'll make it clearer.

Comment (jonatack) [reply test/functional/wallet_avoidreuse.py]: Oh, of course. Yes, I read it wrong.

File: test/functional/wallet_avoidreuse.py (hunk 2 above)
Comment (kallewoof) [reply test/functional/wallet_avoidreuse.py]: Thanks for testing -- I have modified the test to work according to the implementation, which has the limitation that it only marks *new* utxos as used when it sees them -- to get the 100% correct behavior the user would have to run a rescan. To be honest, this should probably be fixed..

File: test/functional/wallet_avoidreuse.py [hunk 3]
```diff
@@ -60,6 +60,7 @@ def assert_unspent(node, total_count=None, total_sum=None, reused_supported=None
 def assert_balances(node, mine):
     '''Make assertions about a node's getbalances output'''
     got = node.getbalances()["mine"]
+    print(got)
```
Comment (jonatack) [inline test/functional/wallet_avoidreuse.py]: Remove this print?

File: test/functional/wallet_avoidreuse.py [hunk 4]
```diff
@@ -240,5 +243,39 @@ def test_fund_send_fund_send(self):
...
+        # getbalances should show no used, 30 btc trusted
+        assert_balances(self.nodes[1], mine={"used": 0, "trusted": 30})
+
+        self.nodes[1].sendtoaddress(retaddr, 11)
+        self.nodes[0].generate(1)
+        self.sync_all()
+
+        # listunspent should show 11 unused outputs
+        # (you might expect 1 unused 9 btc output, and 10 used 1 btc outputs, but due to how things are set up
+        #  this requires a rescan by the user, right now)
```
Comment (jonatack) [inline test/functional/wallet_avoidreuse.py]: If you have to retouch this, suggest:
```
        # listunspent should show 11 unused outputs (you might expect 1 unused 9
        # btc output, and 10 used 1 btc outputs, but due to how things are set
        # up this requires a rescan by the user; this should probably be fixed).
```

**Comment by kallewoof** [comment]
Removed left-over print and tweaked comment.

File: test/functional/wallet_avoidreuse.py [hunk 5]
```diff
@@ -240,5 +242,39 @@ def test_fund_send_fund_send(self):
...
+        Those 10 should all be marked used.
+        A single output of 9 btc should be the change output back to oneself,
+        and should not be marked used.
+        '''
+        self.log.info("Test many utxos one destination")
+
+        fundaddr = self.nodes[1].getnewaddress()
+        retaddr = self.nodes[0].getnewaddress()
+
+        [self.nodes[0].sendtoaddress(fundaddr, 1) for _ in range(30)]
```
Comment (jimmysong) [inline test/functional/wallet_avoidreuse.py]: Generally, this is not a pattern seen in the codebase. When list comprehensions are done, they're done to assign to a variable. It's also a little harder to read. A two-liner like this is easier.:

    for _ in range(30):
        self.nodes[0].sendtoaddress(fundaddr, 1)

Comment (jimmysong) [reply test/functional/wallet_avoidreuse.py]: Also, a comment above would be useful ( `# send one bitcoin to the funding address 30 times`)

**Review by jimmysong — COMMENTED** [review]
Looks good other than 1 nit. Thanks for your PR!

File: test/functional/wallet_avoidreuse.py [hunk 6]
```diff
@@ -240,5 +242,39 @@ def test_fund_send_fund_send(self):
...
+
+        self.nodes[1].sendtoaddress(retaddr, 11)
+        self.nodes[0].generate(1)
+        self.sync_all()
+
+        # listunspent should show 11 unused outputs (you might expect 1 unused 9
+        # btc output, and 10 used 1 btc outputs, but due to how things are set
+        # up this requires a rescan by the user; this should probably be fixed).
+        assert_unspent(self.nodes[1], total_count=11, total_sum=19, reused_supported=True, reused_count=10)
+        assert_balances(self.nodes[1], mine={"used": 0, "trusted": 19})
```
Comment (jonatack) [inline test/functional/wallet_avoidreuse.py]: Verified what happens after a rescan: getbalances changes, but not listunspent.
```diff
+
+        # After -rescan, getbalances shows 1 unused 9 btc output, and 10 used 1 btc outputs.
+        self.restart_node(1, extra_args=self.extra_args[1] + ['-rescan'])
+        assert_unspent(self.nodes[1], total_count=11, total_sum=19, reused_supported=True, reused_count=10)
+        assert_balances(self.nodes[1], mine={"trusted": 9, "used": 10})
```
Worthwhile to add this assertion and update the comment in line 273?

I'd also like to compare and test this on the changes in the related PRs.

File: test/functional/wallet_avoidreuse.py (hunk 5 above)
Comment (kallewoof) [reply test/functional/wallet_avoidreuse.py]: No, there are cases where this is done

Comment (jimmysong) [reply test/functional/wallet_avoidreuse.py]: OK, I missed those. I still don't think it's that clear, but if it's there already, I'm happy to have it be a pattern.

Comment (kallewoof) [reply test/functional/wallet_avoidreuse.py]: Yeah, you're probably right. I'll change it to be more clear and add comment. Thanks for the feedback.

File: test/functional/wallet_avoidreuse.py (hunk 6 above)
Comment (kallewoof) [reply test/functional/wallet_avoidreuse.py]: Good idea! Thanks will add.

**Comment by kallewoof** [comment]
Rendered redundant with #17843.

**Pull request closed without merging**